File: bridges/streams/guest.js
// Canonical source: bridges/streams/guest.js — the stream bridge's JS guest binding (the
// cross-process byte Stream for QuickJS, over the host `__stream_*` ops). Synced into the
// js-runner (crates/rusm-wasm/js-runner/bridge/streams.js) by `make sync-bridges`; edit this
// file, not the copy. `bridge_guest_in_sync` guards drift.
//
// `write` accepts a string (UTF-8) or a Uint8Array; `read` resolves to a Uint8Array or null
// at end-of-stream (host None → undefined → null). The host call still suspends the whole
// instance's fiber (freeing the Tokio worker); the Promise is driven by the QuickJS job
// queue. `Process.openStream`/`acceptStream` (process.js) construct these — so this is
// eval'd before process.js, exposing `Stream` as a global it can reference.

globalThis.Stream = class Stream {
  constructor(handle) {
    this.handle = handle;
  }
  // write accepts a string (UTF-8) or a Uint8Array.
  write(chunk) {
    return typeof chunk === "string"
      ? __stream_write_text(this.handle, chunk)
      : __stream_write(this.handle, chunk);
  }
  close() {
    __stream_close(this.handle);
  }
  // Resolves to a Uint8Array, or null at end-of-stream (host None → undefined → null).
  read() {
    const c = __stream_read(this.handle);
    return Promise.resolve(c === undefined ? null : c);
  }
};

// The RPC layer (rpc.js) tags Stream instances by this reference.
globalThis.__rusm_Stream = globalThis.Stream;

File: bridges/streams/guest.rs
// Canonical source: bridges/streams/guest.rs — the stream bridge's Rust guest binding.
// Synced into rusm-rs (crates/rusm-rs/src/streams.rs) by `make sync-bridges`; edit this
// file, not the copy. The `bridge_guest_in_sync` test fails the build on drift.

//! Cross-process byte streams for a Rust guest — an ergonomic wrapper over the `stream`
//! bridge interface. The write end stays with the opener; the read end is delivered to the
//! target, which [`Stream::accept`]s it. `write`/`read`/`accept` suspend the fiber under
//! back-pressure / until data arrives (the host frees the Tokio worker meanwhile).

// The generated `streams` interface bindings.
use crate::rusm::runtime::streams as abi;
use crate::Pid;

/// A back-pressured byte stream to or from another process. Read suspends until a chunk
/// arrives; `None` is end-of-stream.
pub struct Stream {
    handle: u64,
}

impl Stream {
    /// Open a stream to a pid; `None` if the target is gone.
    pub fn open(to: Pid) -> Option<Stream> {
        abi::stream_open(to.0).map(|handle| Stream { handle })
    }

    /// Block until an incoming stream arrives, and take it for reading.
    pub fn accept() -> Stream {
        Stream {
            handle: abi::stream_accept(),
        }
    }

    /// Write one chunk; `false` once the reader is gone.
    pub fn write(&self, chunk: &[u8]) -> bool {
        abi::stream_write(self.handle, chunk)
    }

    /// Read the next chunk, or `None` at end-of-stream.
    pub fn read(&self) -> Option<Vec<u8>> {
        abi::stream_read(self.handle)
    }

    /// Close the write end (signals end-of-stream to the reader).
    pub fn close(self) {
        abi::stream_close(self.handle);
    }
}
